### PR TITLE
Skip downgrade if requested version below daemon version

### PR DESCRIPTION
**Pull request by nagworld9**
## Description
Skip downgrade if requested version below daemon version.

Issue # 

---

### PR information
- [ ] The title of the PR is clear and informative.
- [ ] There are a small number of commits, each of which has an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [ ] If applicable, the PR references the bug/issue that it fixes in the description.
- [ ] New Unit tests were added for the changes made

### Quality of Code and Contribution Guidelines
- [ ] I have read the [contribution guidelines](https://github.com/Azure/WALinuxAgent/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: azurelinuxagent/ga/agent_update_handler.py [hunk 1]
```diff
@@ -291,6 +300,13 @@ def run(self, goal_state):
             if warn_msg != "":
                 self.__log_event(LogLevel.WARNING, warn_msg)
 
+            daemon_version = self.__get_daemon_version_for_update()
+            if requested_version < daemon_version:
+                # Don't process the update if the requested version is less than daemon version,
```
Comment (narrieta) [inline azurelinuxagent/ga/agent_update_handler.py:305]: let's explain why

Comment (narrieta) [inline azurelinuxagent/ga/agent_update_handler.py:305]: also, let's add a test case for this to agent_update

**Review by narrieta — APPROVED** [review]

File: azurelinuxagent/ga/agent_update_handler.py (hunk 1 above)
Comment (narrieta) [reply azurelinuxagent/ga/agent_update_handler.py]: sorry, I should have been explicit: I meant add this scenario to your end-to-end tests (can be a separate PR). Having a unit test is also great, thanks for adding it.

**Review by narrieta — APPROVED** [review]
Approved with one comment on end-to-end tests

File: azurelinuxagent/ga/agent_update_handler.py (hunk 1 above)
Comment (narrieta) [reply azurelinuxagent/ga/agent_update_handler.py]: never mind, it's there

File: azurelinuxagent/ga/agent_update_handler.py [hunk 2]
```diff
@@ -305,7 +305,7 @@ def run(self, goal_state):
                     # as historically we don't support downgrades below daemon versions. So daemon will not pickup that requested version rather start with
                     # installed latest version again. When that happens agent go into loop of downloading the requested version, exiting and start again with same version.
                     #
-                    raise Exception("Can't process the update as the requested version: {0} is < current daemon version: {1}".format(
+                    raise Exception("as the requested version: {0} is < current daemon version: {1}".format(
```
Comment (narrieta) [inline azurelinuxagent/ga/agent_update_handler.py:308]: Missing "Unable to update Agent:"?

Comment (nagworld9) [reply azurelinuxagent/ga/agent_update_handler.py:308]: we append this msg when we catch the exception.

Comment (narrieta) [reply azurelinuxagent/ga/agent_update_handler.py:308]: That is confusing. Myabe instead define an AgentUpdateError exception with a "reason" property and initialize the reason to "Requested version... is < ...."? The format the message where you handle the exception.

Comment (narrieta) [reply azurelinuxagent/ga/agent_update_handler.py:308]: Or actually, you don't need the "reason" property. The type of the exception would be AgentUpdateError and the message  "Requested version... is < ...."

Comment (nagworld9) [reply azurelinuxagent/ga/agent_update_handler.py]: updated

**Review by narrieta — APPROVED** [review]

**Pull request merged**